include/ocpp/v201/smart_charging.hpp

```diff
@@ -85,6 +106,40 @@ class SmartChargingHandler {
     ///
     std::vector<ChargingProfile> get_profiles();
 
+    ///
+    /// \brief Iterates over the periods of the given \p profile and returns a struct that contains the period and the
+    /// absolute end time of the period that refers to the given absoulte \p time as a pair.
+    ///
+    PeriodDateTimePair find_period_at(const ocpp::DateTime& time, const ChargingProfile& profile, const int evse);
+
+    ///
+    /// \brief Gets the absolute start time of the given \p profile for the given \p connector_id for different profile
+    /// purposes
+    ///
+    std::optional<ocpp::DateTime> get_profile_start_time(const ChargingProfile& profile, const ocpp::DateTime& time,
+                                                         const int evse);
+
+    ///
+    /// \brief Calculates the composite schedule for the given \p valid_profiles and the given \p connector_id
+    ///
+    CompositeSchedule calculate_composite_schedule(std::vector<ChargingProfile> valid_profiles,
+                                                   const ocpp::DateTime& start_time, const ocpp::DateTime& end_time,
+                                                   const int32_t evse_id, ChargingRateUnitEnum charging_rate_unit);
+
+    static int32_t determine_duration(const ocpp::DateTime& start_time, const ocpp::DateTime& end_time);
+
+    static bool within_time_window(const ocpp::DateTime& start_time, const ocpp::DateTime& end_time);
+
+    ///
+    /// \brief Iterates over the periods of the given \p valid_profiles and determines the earliest next absolute period
+    /// end time later than \p temp_time
+    ///
+    ocpp::DateTime get_next_temp_time(const ocpp::DateTime temp_time,
+                                      const std::vector<ChargingProfile>& valid_profiles, const int32_t evse_id);
+
+    ocpp::DateTime get_period_end_time(const int period_index, const ocpp::DateTime& period_start_time,
+                                       const ChargingSchedule& schedule);
+
```

Comment:
Based on previous feedback anything that is not `calculate_composite_schedule` EVerest will want to have it `protected` if it is not being directly used.

Reply:
'Tis done.